File: dynamic_listings/javascript/default/collisiondetection/pixel_perfect.js
class Color{
    constructor(){
        this.colorData = [];
    }

    isWhite(){
        // ...
    };
}

class Bitmask{
    constructor(){
        this.data = [];  // An array of Color() classes
    }
    getColor(x, y){
        // ...
    };
    // ...
}

class Sprite{
    constructor(){
        this.bitmask = new Bitmask();
        this.x = null;
        this.y = null;
        this.width = 0;
        this.height = 0;
    }
}

function pixel_perfect_collision(A, B){
    // Calculate the intersecting rectangle to limit checks
    let x1 = max(A.x, B.x);
    let x2 = min((A.x + A.width), (B.x + B.width));

    let y1 = max(A.y, B.y);
    let y2 = min((A.y + A.height), (B.y + B.height));

    // For each pixel in the intersecting rectangle, let's check
    for (let y = y1; y <= y2; y++){
        for (let x = x1; x <= x2; x++){
            // We're working in the intersecting triangle, so we'll need to
            // rework our coordinates
            let a = A.bitmask.getColor(x - A.x, y - A.y);
            let b = B.bitmask.getColor(x - B.x, y - B.y);

            if (a.isWhite() && b.isWhite()){
                return true;
            }
        }
    }

    // If no collision is occurred by the end of the checking, we're safe
    return false;
}

File: dynamic_listings/C++/default/collisiondetection/pixel_perfect.cpp
#include<algorithm>
struct Color{
    int colorData;
    bool isWhite{
        //...
    }
};

struct Bitmask{
    Color data[];
    Color getColor(int x, int y){
        // ...
    }
};

struct Sprite{
    Bitmask bitmask;
    int x;
    int y;
    int width;
    int height;
};


bool pixel_perfect_collision(Sprite A, Sprite B){
    // Calculate the intersecting rectangle to limit checks
    int x1 = std::max(A.x, B.x);
    int x2 = std::min((A.x + A.width), (B.x + B.width));

    int y1 = std::max(A.y, B.y);
    int y2 = std::min((A.y + A.height), (B.y + B.height));

    // For each pixel in the intersecting rectangle, let's check
    for (int y=y1; y < y2; y++){
        for (int x=x1; x < x2; x++){
            // We're working in the intersecting triangle, so we'll need to
            // rework our coordinates
            Color a = A.bitmask.getColor(x - A.x, y - A.y);
            Color b = B.bitmask.getColor(x - B.x, y - B.y);
            if (a.isWhite() && b.isWhite()){
                return true;
            }
        }
    }

    // If no collision is occurred by the end of the checking, we're safe
    return false;
}
